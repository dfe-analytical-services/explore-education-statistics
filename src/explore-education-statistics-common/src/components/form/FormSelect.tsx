import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import { Dictionary } from '@common/types';
import classNames from 'classnames';
import orderBy from 'lodash/orderBy';
import React, {
  ChangeEventHandler,
  CSSProperties,
  FocusEventHandler,
  ReactNode,
} from 'react';
import ErrorMessage from '../ErrorMessage';

export type SelectChangeEventHandler = ChangeEventHandler<HTMLSelectElement>;

export interface SelectOption<Value = string | number> {
  label: string;
  value: Value;
  selected?: boolean;
  style?: CSSProperties;
}

export interface FormSelectProps extends FormLabelProps {
  className?: string;
  disabled?: boolean;
  error?: string;
  id: string;
  inline?: boolean;
  hint?: string;
  name: string;
  onBlur?: FocusEventHandler;
  onChange?: SelectChangeEventHandler;
  options?: SelectOption[];
  optGroups?: Dictionary<SelectOption[]>;
  order?:
    | (keyof SelectOption)[]
    | ((option: SelectOption) => SelectOption[keyof SelectOption])[];
  orderDirection?: ('asc' | 'desc')[];
  placeholder?: string;
  value?: string;
}

const FormSelect = ({
  className,
  disabled,
  error,
  id,
  inline = false,
  hint,
  hideLabel,
  label,
  name,
  onBlur,
  onChange,
  options,
  optGroups,
  order = ['label'],
  orderDirection = ['asc'],
  placeholder,
  value,
}: FormSelectProps) => {
  return (
    <>
      <FormSelectWrapper inline={inline}>
        <FormLabel
          className={classNames({ 'govuk-!-margin-right-2': inline })}
          id={id}
          label={label}
          hideLabel={hideLabel}
        />

        {/* Hint and error moved below the select when inline */}
        {!inline && (
          <>
            {hint && (
              <span id={`${id}-hint`} className="govuk-hint">
                {hint}
              </span>
            )}
            {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
          </>
        )}

        <select
          aria-describedby={
            classNames({
              [`${id}-error`]: !!error,
              [`${id}-hint`]: !!hint,
            }) || undefined
          }
          className={classNames('govuk-select', className, {
            'govuk-select--error': !!error,
          })}
          id={id}
          name={name}
          disabled={disabled}
          onBlur={onBlur}
          onChange={onChange}
          value={value}
        >
          {placeholder && <option value="">{placeholder}</option>}
          {options &&
            (order === undefined || order.length === 0
              ? options
              : orderBy(options, order, orderDirection)
            ).map(option => (
              <option
                value={option.value}
                key={`${option.value}-${option.label}`}
                selected={option.selected}
                style={option.style}
              >
                {option.label}
              </option>
            ))}
          {optGroups &&
            Object.keys(optGroups)
              .sort()
              .map(group => (
                <optgroup key={group} label={group}>
                  {optGroups[group].map(option => (
                    <option
                      key={`value-${option.value}`}
                      value={option.value}
                      style={option.style}
                    >
                      {option.label}
                    </option>
                  ))}
                </optgroup>
              ))}
        </select>
      </FormSelectWrapper>
      {inline && (
        <>
          {hint && (
            <span id={`${id}-hint`} className="govuk-hint">
              {hint}
            </span>
          )}
          {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
        </>
      )}
    </>
  );
};

FormSelect.unordered = [] as [];

export default FormSelect;

function FormSelectWrapper({
  inline,
  children,
}: {
  inline: boolean;
  children: ReactNode;
}) {
  return inline ? (
    <div className="dfe-flex dfe-align-items--center">{children}</div>
  ) : (
    // eslint-disable-next-line react/jsx-no-useless-fragment
    <>{children}</>
  );
}
