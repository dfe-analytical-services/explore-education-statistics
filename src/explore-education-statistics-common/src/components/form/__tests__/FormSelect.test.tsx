import React from 'react';
import { render } from '@testing-library/react';
import FormSelect from '../FormSelect';

describe('FormSelect', () => {
  test('renders correctly with required props', () => {
    const { container, getByLabelText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-1', label: 'Option 1' },
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
        ]}
      />,
    );

    expect(getByLabelText('Test select')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders list of options in alphabetical order by default', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
          { value: 'option-1', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 1');
    expect(options[1]).toHaveTextContent('Option 2');
    expect(options[2]).toHaveTextContent('Option 3');
  });

  test('renders list of options in reverse alphabetical order', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        orderDirection={['desc']}
        options={[
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
          { value: 'option-1', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 3');
    expect(options[1]).toHaveTextContent('Option 2');
    expect(options[2]).toHaveTextContent('Option 1');
  });

  test('renders list of options in custom order', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        order={['value']}
        options={[
          { value: '1', label: 'Option 2' },
          { value: '2', label: 'Option 3' },
          { value: '3', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 2');
    expect(options[1]).toHaveTextContent('Option 3');
    expect(options[2]).toHaveTextContent('Option 1');
  });

  test('renders with error message', () => {
    const { container, getByText, getByLabelText } = render(
      <FormSelect
        id="test-select"
        error="Something went wrong"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-1', label: 'Option 1' },
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
        ]}
      />,
    );

    expect(getByText('Something went wrong')).toHaveClass(
      'govuk-error-message',
    );
    expect(getByLabelText('Test select')).toHaveClass('govuk-select--error');
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('aria-describedby is equal to the hint id', () => {
    const { getByLabelText, getByText } = render(
      <FormSelect
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        options={[{ value: 'option-1', label: 'Option 1' }]}
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-hint',
    );
  });

  test('aria-describedby is equal to the error id', () => {
    const { getByLabelText, getByText } = render(
      <FormSelect
        id="test-input"
        label="Test input"
        name="testInput"
        error="Field is required"
        options={[{ value: 'option-1', label: 'Option 1' }]}
      />,
    );

    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-error',
    );
  });

  test('aria-describedby contains both hint and error ids', () => {
    const { getByLabelText, getByText } = render(
      <FormSelect
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        error="Field is required"
        options={[{ value: 'option-1', label: 'Option 1' }]}
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );

    const ariaDescribedBy = getByLabelText('Test input').getAttribute(
      'aria-describedby',
    );

    expect(ariaDescribedBy).toContain('test-input-error');
    expect(ariaDescribedBy).toContain('test-input-hint');
  });
});
