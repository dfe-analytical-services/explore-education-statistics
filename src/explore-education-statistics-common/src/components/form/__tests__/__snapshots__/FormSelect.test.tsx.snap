// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect renders correctly with required props 1`] = `
<label class="govuk-label"
       for="test-select"
>
  Test select
</label>
<select class="govuk-select"
        id="test-select"
        name="testSelect"
>
  <option value="option-1">
    Option 1
  </option>
  <option value="option-2">
    Option 2
  </option>
  <option value="option-3">
    Option 3
  </option>
</select>
`;

exports[`FormSelect renders with error message 1`] = `
<label class="govuk-label"
       for="test-select"
>
  Test select
</label>
<span class="govuk-error-message"
      id="test-select-error"
>
  Something went wrong
</span>
<select aria-describedby="test-select-error"
        class="govuk-select govuk-select--error"
        id="test-select"
        name="testSelect"
>
  <option value="option-1">
    Option 1
  </option>
  <option value="option-2">
    Option 2
  </option>
  <option value="option-3">
    Option 3
  </option>
</select>
`;
